src/Parser/Types.hs

```diff
@@ -227,19 +219,38 @@ tyBotP = do
 
 -- | Parse atomic types (i,e, without tyop chains)
 typAtomP :: Parser (Typ, SourcePos)
-typAtomP = tyParensP
-  <|> nominalTypeP
-  <|> xdataOrRefinementP Data
-  <|> xdataOrRefinementP Codata
-  <|> recTypeP
-  <|> tyTopP
-  <|> tyBotP
-  <|> tyI64P
-  <|> tyF64P
-  <|> tyCharP
-  <|> tyStringP
-  <|> typeVariableP
+typAtomP = do 
+  startPos <- getSourcePos
+  (fstTy, _) <- 
+    tyParensP
+    <|> nominalTypeP
+    <|> xdataOrRefinementP Data
+    <|> xdataOrRefinementP Codata
+    <|> recTypeP
+    <|> tyTopP
+    <|> tyBotP
+    <|> tyI64P
+    <|> tyF64P
+    <|> tyCharP
+    <|> tyStringP
+    <|> typeVariableP
+  args <- optional tyArgsP
+  endPos <- getSourcePos
+  case args of
+    Nothing -> pure (fstTy, endPos)
+    Just (args',endPos) -> pure (TyApp (Loc startPos endPos) fstTy args',endPos)
 
+tyArgsP :: Parser (NonEmpty Typ, SourcePos)
+tyArgsP = do 
+  symbolP SymParenLeft 
+  sc
+  args <- typP `sepBy1` (symbolP SymComma >> sc)
+  symbolP SymParenRight
+  sc
```

Comment:
I think you have to change the order of `sc` and `getSourcePos`, since the endPos is the last non-space character being parsed.